spec/ics-005-port-allocation/README.md

```diff
@@ -0,0 +1,178 @@
+which can be passed to a module and will not be duplicable by other modules. Two examples are store keys as used in the Cosmos SDK ([reference](https://github.com/cosmos/cosmos-sdk/blob/master/store/types/store.go#L224))
+and object references as used in Agoric's Javascript runtime ([reference](https://github.com/Agoric/SwingSet)).
+
+```typescript
+type CapabilityKey object
+```
+
+```typescript
+function newCapabilityKey(): CapabilityKey {
+  // provided by host state machine, e.g. pointer address in Cosmos SDK
+}
+```
+
+In the latter case, the IBC handler must have the ability to securely read the *source identifier* of the calling module,
+a unique string for each module in the host state machine, which cannot be altered by the module or faked by another module.
+An example is smart contract addresses as used by Ethereum ([reference](https://ethereum.github.io/yellowpaper/paper.pdf)).
+
+```typescript
+type SourceIdentifier string
+```
+
+```typescript
+function callingModuleIdentifier(): SourceIdentifier {
+  // provided by host state machine, e.g. contract address in Ethereum
+}
+```
+
+`generate` and `authenticate` functions are then defined as follows.
+
+In the former case, `generate` returns a new object-capability key, which must be returned by the outer-layer function, and `authenticate` requires that the outer-layer function take an extra argument `capability`, which is an object-capability key with uniqueness enforced by the host state machine.
+
+```
+function generate(): CapabilityKey {
+  return newCapabilityKey()
+}
+```
+
+```
+function authenticate(key: CapabilityKey) {
+  return capability === key
+}
+```
+
+In the latter case, `generate` returns the calling module's identifier and `authenticate` merely checks it.
+
+```
+function generate(): SourceIdentifier {
+  return callingModuleIdentifier()
+}
+```
+
+```
+function authenticate(id: SourceIdentifier) {
+  return callingModuleIdentifier() === id
+}
+```
+
+### Subprotocols
+
+#### Preliminaries
+
+`portKey` takes an `Identifier` and returns the store key under which the object-capability reference or owner module identifier associated with a port should be stored.
+
+```typescript
+function portKey(id: Identifier) {
+  return "ports/{id}"
+}
+```
+
+#### Binding to a port
+
+The IBC handler MUST implement `bindPort`. `bindPort` binds to an unallocated port, failing if the port has already been allocated.
+
+```typescript
+function bindPort(id: Identifier) {
+  assert(get(portKey(id)) === null)
+  key = generate()
+  set(key, portKey(id))
+  return key
+}
+```
+
+#### Transferring ownership of a port
+
+If the host state machine supports object-capability keys, no additional protocol is necessary, since the port reference is a bearer capability. If it does not, the IBC handler MAY implement the following `transferPort` function.
+
+```typescript
+function transferPort(id: Identifier) {
+  assert(authenticate(get(portKey(id))))
```

Comment:
should it also release the port before assigning a new value to it `del(portKey(id))`?